### PR TITLE
osx build with unittest

### PR DESCRIPTION
**Pull request by gyanesh-m**
Trial to fix build for osx.
Fixes : #1901

### CODE REVIEW

**Comment by menshikh-iv** [comment]
@gyanesh-m hey, great, you reproduced a problem (with stuck of d2v), now you need to investigate, what's exactly happens and why.

**Comment by gyanesh-m** [comment]
@menshikh-iv Yea sure , I will get started with it. Also is it advisable to edit .travis.yml file manually even though it is generated automatically ?

**Comment by gyanesh-m** [comment]
@menshikh-iv Hey, its been 5 hours but job in travis-ci is not getting scheduled. Any ideas ?

**Comment by menshikh-iv** [comment]
@gyanesh-m sometimes Travis stuck (or have no free `MacOSX` machines), this is the standard situation.
I re-run your build, I hope that this will help.

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      source /Users/travis/miniconda3/bin/activate root
+      conda config --remove channels defaults
+      conda config --add channels defaults
+      conda config --add channels conda-forge
+      conda config --set show_channel_urls true
+      conda install --yes --quiet conda-forge-build-setup
+      source run_conda_forge_build_setup
+
+script:
+  - travis_wait conda build ./recipe
```
Comment (menshikh-iv) [inline .travis.yml]: this doesn't look like the solution, but you can try (in debugging propose).

**Comment by souravsingh** [comment]
@gyanesh-m conda-forge can have problems in requisitioning machines when many packages run through CI services. I suggest waiting  for a while.

**Comment by souravsingh** [comment]
Travis Wait command fails. I would suggest looking for other solutions.

**Comment by menshikh-iv** [comment]
I propose
- check "FAST_VERSION" before run tests (should be `!= -1`)
- run ONLY doc2vec tests with max verbosity (`logging.DEBUG`)

**Comment by menshikh-iv** [comment]
@souravsingh good news, I added `SKIP_NETWORK_TESTS` flag (for cases when CI have some problems with SSL or any network issues).

**Comment by menshikh-iv** [comment]
@gyanesh-m not quiet, when you run it in current manner - no debug logs, for more pricely testing - `python -m unittest -v gensim.test.<test_file>.<class_name>.<method_name>`

**Comment by gyanesh-m** [comment]
@menshikh-iv I have written the test file for fast version check, so should I add it here or in the main repo ?

**Comment by menshikh-iv** [comment]
@gyanesh-m you can check it in one line (no need to add it as distinct test to main repo)

**Comment by gyanesh-m** [comment]
@menshikh-iv In the travis build log I can't find the line which shows the test for fast version check mentioned in test_fast.py. Any idea why it didn't execute ?

**Comment by menshikh-iv** [comment]
@gyanesh-m maybe you should modify https://github.com/conda-forge/gensim-feedstock/blob/2f18b2fca0e96457e8b67e060de3c3448891bd12/recipe/forge_test.sh ?

**Comment by menshikh-iv** [comment]
@gyanesh-m or this one https://github.com/conda-forge/gensim-feedstock/pull/12/files#diff-e178b687b10a71a3348107ae3154e44cR63 (maybe this line is stuck and you didn't achieve your line)

**Comment by gyanesh-m** [comment]
@menshikh-iv Oh ya, got it. It needs to be added in forge_test.sh. Thanks.

**Comment by souravsingh** [comment]
@gyanesh-m You seem to be adding un-necessary commands into the test section. Refer to this- https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html

**Comment by menshikh-iv** [comment]
@souravsingh I hope that this is only for debug proposes (and will be removed after of course)

**Comment by gyanesh-m** [comment]
@souravsingh Hi, don't worry these are for debugging purposes only until I find the bug. They won't be seen in the final commit.

**Comment by gyanesh-m** [comment]
@menshikh-iv Hey, can you restart the travis-ci build. I think it's stuck since its been more than 12 hours.

**Comment by menshikh-iv** [comment]
@gyanesh-m done, you can also do it yourself with close & open current PR.

**Comment by gyanesh-m** [comment]
@menshikh-iv @souravsingh Hi, the logs show that the test does complete but takes around 22 minutes to do so. However, since I have my exams in the coming week I won't be able to do much for sometime now. I will resume this task next Sunday.

**Comment by menshikh-iv** [comment]
@gyanesh-m OK, thanks for the information, what's about "FAST_VERSION"?

**Comment by gyanesh-m** [comment]
@menshikh-iv FAST_VERSION test passed successfully.

**Comment by menshikh-iv** [comment]
@gyanesh-m oh, this is really wired (what's exact value of `FAST_VERSION`?)

**Comment by gyanesh-m** [comment]
@menshikh-iv Actually in the test I just checked for the condition that FAST_VERSION shouldn't equal to -1. I will rerun the test now to display the FAST_VERSION value and will let you know when it's done.

**Comment by gyanesh-m** [comment]
@menshikh-iv The test in travis-ci for python versions 2.7, 3.5 and 3.6 all gave the value of FAST_VERSION as 1.

**Comment by souravsingh** [comment]
We can ask one of the people from conda-forge for the solution. One solution I can think of is running a small subset of tests which are important(something which numba does).

**Comment by gyanesh-m** [comment]
@menshikh-iv I checked the travis-ci logs and I can't find test for fasttext or word2vec. This is because they were supposed to run after doc2vec but couldn't as it terminated early. What I can do is run the test for word2vec and fasttext both separately as I did for doc2vec so that we can understand the bug better.
Also the gensim-wheels which you mentioned, did that also run on Travis ? If so can't we modify the configuration file of it to suit our need ?
Also, I was thinking of cross checking the debug logs later from travis-ci against the ones which I run locally to get an idea of which function/module is consuming time.

**Comment by gyanesh-m** [comment]
@souravsingh Shouldn't we keep that as a last resort since it will affect the coverage of the build ?

**Comment by menshikh-iv** [comment]
@souravsingh yes, but now the goal is to understand what happens / why / how to fix it. 
Otherwise, we would have already ignore this test and add MacOSX.

**Comment by menshikh-iv** [comment]
@gyanesh-m about gensim-wheels - yes, MacOSX on Travis (but this use `multibuild` toolkit, don't involve `multibuild` here). Unfortunately, this is too different to the comparison.

**Comment by gyanesh-m** [comment]
@menshikh-iv Hey, I just checked the logs and what I found was fasttext test took around 3 minutes for 2.7, 3.5 and 3.6 python environment. But word2vec failed to run for 2.7 and took around 27 minutes for 3.5 and 3.6 python environment. Also, I can't output the complete debug log for word2vec in travis-ci as in the previous commit( https://github.com/conda-forge/gensim-feedstock/pull/12/commits/21bb7c0d4981642d61aeef761163c1d556a03a76)  the debug output over flooded the log file and the session couldn't continue as it reached max log file size of 4 MB. So, does travis-ci provides an option to increase log file size ?

**Comment by menshikh-iv** [comment]
@gyanesh-m look to https://github.com/travis-ci/travis-ci/issues/1382 https://github.com/travis-ci/docs-travis-ci-com/issues/281 https://github.com/travis-ci/travis-ci/issues/1113 https://stackoverflow.com/questions/26082444/how-to-work-around-travis-cis-4mb-output-limit

**Comment by gyanesh-m** [comment]
@menshikh-iv There is a timeout in the travis build as the job is taking longer than 50 minutes. Because of this it is not able to run completely. Is 50 minutes the maximum limit or can it be altered ?

**Comment by menshikh-iv** [comment]
@gyanesh-m I don't know (google it please), anyway, this is unacceptable time (should be in interval 18-35 minutes)

**Comment by gyanesh-m** [comment]
@menshikh-iv Hi, I forgot to mention the results for the last run. It took word2vec around 27 mins.  I ran the test only for word2vec and fast version value was 1. This is the similar behavior which was observed in doc2vec test. I was thinking maybe we can use profiling to give us concrete idea which function call is consuming more time. 
P.S. I will resume this once I am done with my proposal.

**Comment by menshikh-iv** [comment]
@gyanesh-m can we conclude that all models with `cython` works really slow (but code compiled correctly)?

**Comment by gyanesh-m** [comment]
@menshikh-iv Yes, from current findings that seems correct.

**Comment by gyanesh-m** [comment]
@menshikh-iv Hi, earlier I tried to profile the test cases with pytest-profiling but I was facing some difficulty in its installation as I couldn't find it in any existing conda-forge channel. So I did the profiling using cprofile for word2vec test file and what I found was the method which deals with acquiring of thread locks in the threading library executes 30 times more slower in osx than in linux. This leads to the completion of test file in linux in around 55s but the same test takes around 26 minutes in osx. Also, the test file in osx makes around 1300 more calls to the threading library as compared to the linux version. The graphs and the corresponding stat files are available [here](https://drive.google.com/drive/folders/1utnnnQi4K1dnor_ixC6EdLDYsh9ZnXMJ?usp=sharing).

**Comment by menshikh-iv** [comment]
Thanks for the investigation @gyanesh-m, looks like optimized version doesn't work on OSX =/

**Comment by gyanesh-m** [comment]
@menshikh-iv I looked up and found that multi threaded program in mac osx can be slower than single threaded for the BLAS library provided by Apple as compared to the one available from MacPorts[ [Source] ](https://stackoverflow.com/questions/19606276/multi-threaded-program-not-faster-than-single-thread-on-mac-but-on-linux-unles). I am not sure if this is true but this can be tried as a solution.

**Comment by menshikh-iv** [comment]
@gyanesh-m one important tip - we build OSX wheels for PyPI for a long time and this works as expected, check https://github.com/MacPython/gensim-wheels/.

Also, please check how works https://github.com/conda-forge/scikit-learn-feedstock (and numpy/scipy), I think we missed something critical here.

**Comment by gyanesh-m** [comment]
@menshikh-iv Hey, I haven't looked into this for quite a while and I am busy for sometime now, so I won't be able to help further in this.

**Comment by menshikh-iv** [comment]
@gyanesh-m ok, but can you describe please, what are you tried, what you found, this will be really useful for future investigation, thanks!

**Comment by gyanesh-m** [comment]
@menshikh-iv Ok, I will do it over the weekend.

**Comment by menshikh-iv** [comment]
outdated, closing.

**Pull request closed without merging**